docs(speech_recording): feedback api draft

dcsiebe opened the pull request:
Add a first draft of the feedback api for speech recordings.

dcsiebe committed:
docs(speech_recording): feedback api draft

Add a first draft of the feedback api for speech recordings.

joggienl reviewed (changes requested):
This way getting the feedback, it is very tightly coupled to the existing Speech Recording. I thought we wanted to separate it a bit, because of its premium nature?

Also, I can't get the "dance of the RPC's" here. What do I need to execute in order to get feedback? Just feedback?

I asume:
1. `nl.itslanguage.recording.init_recording`
1. `nl.itslanguage.recording.init_challenge`
1. `nl.itslanguage.recording.feedback`

I'm not sure where the actual streaming part kicks in here. Also the "older" `nl.itslanguage.recording.write`, `nl.itslanguage.recording.close` and `nl.itslanguage.recording.init_audio` are still in this document. Those are no more needed when we step to streaming?

kvdb commented on docs/websocket/speech_recording.md:
```diff
@@ -106,3 +106,63 @@ The RPC can return various errors:
+Name     | Type  | Description
+---------|-------|------------
+sentence | `int` | The index of the sentence, starting at `0`.
+feedback | `int` | Feedback score, `0` means bad and `1` means good.
```
perhaps use a bool: recognitionAccepted: true/false

dcsiebe replied:
CLD returns a float as feedback (same as here). I figured that it might in the future indicate how well something was pronounced, like `0.6` for an okay pronunciation. If this is the case a bool doesn't suffice.
Any thoughts on this?

joggienl replied:
Well, bool makes more sense then in terms of incremental development.

Feedback will be more than a float because it will be more than just bad/good and an percentage. At least, that is the required situation.

kvdb reviewed (commented):
furthermore I think Jogchums comments make sense.
I'm also wondering how to start recording at the 3th sentence when frontend decides the user needs to re-speak that sentence. To make that work start/stop rpc's should be described as well. perhaps re-using init_recording(int sentenence)

dcsiebe commented:
@joggienl 
It it a separate call, so I think it's decoupled enough. We have the option of re-using a lot of stuff from the recordings, so I'm thinking; why not? If the feedback rpc is only available for premium users the feature is shielded enough imo.

The overview of rpc's needed (a.k.a. the dance) is missing indeed, I'll add such an overview. ;)

The older `init_audio`, `write` and `close` rpcs are still needed for the recording. The feedback rpc can utilize streaming in the new way. Refactoring the old rpcs to a streaming format is out of the scope of this pr.

dcsiebe commented:
@kvdb That's a valid point, I overlooked that part. I'll make some changes to support that.

joggienl commented:
> The overview of rpc's needed (a.k.a. the dance) is missing indeed, I'll add such an overview. ;)

Thanks!

> It it a separate call, so I think it's decoupled enough. We have the option of re-using a lot of stuff from the recordings, so I'm thinking; why not? If the feedback rpc is only available for premium users the feature is shielded enough imo.

and

> The older init_audio, write and close rpcs are still needed for the recording. The feedback rpc can utilize streaming in the new way. Refactoring the old rpcs to a streaming format is out of the scope of this pr.

So in effect, I can just use the feedback to do the recording at the same time? if so, I would move it out to another namespace (or however you would like to call it). This way in effect you offer two ways of recording and that will be confusing.

@kvdb do you have a specific view on this?

dcsiebe committed:
docs(speech_recording): add rpc overview for feedback

Describe which calls are needed in order to perform a recording with
feedback.

dcsiebe commented:
After offline discussion it's clear that the feedback stuff should reside on a separate page. This will cause some duplication but they may grow differently later on.

dcsiebe committed:
docs(feedback): define in separate page

Define the feedback rpc in a separate page on a separate base URI.

dcsiebe commented:
FYI pause/resume is yet to be done so keep this pr open ;)

joggienl commented:
I like the new changes 👍

dcsiebe committed:
docs(feedback): add pause/resume

Add docs for a pause and resume rpc so feedback recordings can be paused
and resumed.

dcsiebe commented:
@kvdb I think that adheres to the discussion of last week (design the api for usability or with less technical roadblocks).  I believe that sentence makes more sense to the user, which is important imo.
@joggienl Do you have an opinion to throw in the mix?

joggienl commented on docs/websocket/feedback.md:
```diff
@@ -0,0 +1,131 @@
+
+Name        | Type     | Description
+------------|----------|------------
+id          | `string` | **Required** The id of the feedback to pause.
```
I think you mean `resume` instead of `pause`

dcsiebe replied:
Good catch!

joggienl commented:
I think that both could be possible, word and sentence I mean.
For now, sentence is good. But we could add word into the mix later on, not now.

Counting words to start at a sentence is weird, and visa versa.

I think both approaches are equal in technical terms to the end users, "word" and "sentence" are both related to text and don't throw technical roadblocks imo.

dcsiebe committed:
docs(feedback): fix copy/paste fail

In the resume rpc "pause" is named where "resume" applies.

joggienl commented:
Maybe it is good to include in the documentation as well, if you call the `listend_and_reply` rpc it won't finish until you finish the registered RCP call.

So, if you call `pause` it does not end the `listend_and_reply` call. Only the `rpc` you call with the `listen_and_reply` can.

dcsiebe committed:
docs(feedback): emphasize stop audio processing

Add a note on how audio processing and feedback is stopped. Also add a
note that emphasizes that pausing does not mean stopping.
